{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Thoth: Build Analysers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Introduction\" data-toc-modified-id=\"Introduction-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Introduction</a></span><ul class=\"toc-item\"><li><span><a href=\"#Lead-by-Example:\" data-toc-modified-id=\"Lead-by-Example:-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Lead by Example:</a></span></li><li><span><a href=\"#How-it-works\" data-toc-modified-id=\"How-it-works-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>How it works</a></span><ul class=\"toc-item\"><li><span><a href=\"#The-intuition\" data-toc-modified-id=\"The-intuition-1.2.1\"><span class=\"toc-item-num\">1.2.1&nbsp;&nbsp;</span>The intuition</a></span></li></ul></li></ul></li><li><span><a href=\"#Installation\" data-toc-modified-id=\"Installation-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Installation</a></span></li><li><span><a href=\"#Local-setup\" data-toc-modified-id=\"Local-setup-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Local setup</a></span></li><li><span><a href=\"#Report\" data-toc-modified-id=\"Report-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Report</a></span></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "This is a set of analysers that will retrieve information from build logs.\n",
    "\n",
    "<br>\n",
    "\n",
    "### Lead by Example:\n",
    "\n",
    "Consider a log produced by the [pip](https://pypi.org/project/pip/) package installer:\n",
    "\n",
    "> We can also use [pipenv](https://pypi.org/project/pipenv/), see the [How it works](how-it-works) section\n",
    "\n",
    "<details>\n",
    "  <summary> Log file content </summary>\n",
    "  <pre>\n",
    "Processing /home/macermak/RedHat/aicoe/lab\n",
    "Collecting networkx (from thoth-lab==0.0.3)\n",
    "Collecting requests (from thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/7d/e3/20f3d364d6c8e5d2353c72a67778eb189176f08e873c9900e10c0287b84b/requests-2.21.0-py2.py3-none-any.whl\n",
    "Collecting pandas (from thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/e6/de/a0d3defd8f338eaf53ef716e40ef6d6c277c35d50e09b586e170169cdf0d/pandas-0.24.1-cp36-cp36m-manylinux1_x86_64.whl\n",
    "Collecting plotly (from thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/fd/db/003b5cfbc710f4d4982440451185b952269e4080a57ae7e760a2ceb8ce0c/plotly-3.6.1-py2.py3-none-any.whl\n",
    "Collecting decorator>=4.3.0 (from networkx->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/f1/cd/7c8240007e9716b14679bc217a1baefa4432aa30394f7e2ec40a52b1a708/decorator-4.3.2-py2.py3-none-any.whl\n",
    "Collecting idna<2.9,>=2.5 (from requests->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/14/2c/cd551d81dbe15200be1cf41cd03869a46fe7226e7450af7a6545bfc474c9/idna-2.8-py2.py3-none-any.whl\n",
    "Collecting urllib3<1.25,>=1.21.1 (from requests->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/62/00/ee1d7de624db8ba7090d1226aebefab96a2c71cd5cfa7629d6ad3f61b79e/urllib3-1.24.1-py2.py3-none-any.whl\n",
    "Collecting certifi>=2017.4.17 (from requests->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/9f/e0/accfc1b56b57e9750eba272e24c4dddeac86852c2bebd1236674d7887e8a/certifi-2018.11.29-py2.py3-none-any.whl\n",
    "Collecting chardet<3.1.0,>=3.0.2 (from requests->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/bc/a9/01ffebfb562e4274b6487b4bb1ddec7ca55ec7510b22e4c51f14098443b8/chardet-3.0.4-py2.py3-none-any.whl\n",
    "Collecting python-dateutil>=2.5.0 (from pandas->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/41/17/c62faccbfbd163c7f57f3844689e3a78bae1f403648a6afb1d0866d87fbb/python_dateutil-2.8.0-py2.py3-none-any.whl\n",
    "Collecting pytz>=2011k (from pandas->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/61/28/1d3920e4d1d50b19bc5d24398a7cd85cc7b9a75a490570d5a30c57622d34/pytz-2018.9-py2.py3-none-any.whl\n",
    "Collecting numpy>=1.12.0 (from pandas->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/f5/bf/4981bcbee43934f0adb8f764a1e70ab0ee5a448f6505bd04a87a2fda2a8b/numpy-1.16.1-cp36-cp36m-manylinux1_x86_64.whl\n",
    "Collecting six (from plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl\n",
    "Collecting retrying>=1.3.3 (from plotly->thoth-lab==0.0.3)\n",
    "Collecting nbformat>=4.2 (from plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/da/27/9a654d2b6cc1eaa517d1c5a4405166c7f6d72f04f6e7eea41855fe808a46/nbformat-4.4.0-py2.py3-none-any.whl\n",
    "Collecting traitlets>=4.1 (from nbformat>=4.2->plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/93/d6/abcb22de61d78e2fc3959c964628a5771e47e7cc60d53e9342e21ed6cc9a/traitlets-4.3.2-py2.py3-none-any.whl\n",
    "Collecting jupyter-core (from nbformat>=4.2->plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/1d/44/065d2d7bae7bebc06f1dd70d23c36da8c50c0f08b4236716743d706762a8/jupyter_core-4.4.0-py2.py3-none-any.whl\n",
    "Collecting jsonschema!=2.5.0,>=2.4 (from nbformat>=4.2->plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/77/de/47e35a97b2b05c2fadbec67d44cfcdcd09b8086951b331d82de90d2912da/jsonschema-2.6.0-py2.py3-none-any.whl\n",
    "Collecting ipython-genutils (from nbformat>=4.2->plotly->thoth-lab==0.0.3)\n",
    "  Using cached https://files.pythonhosted.org/packages/fa/bc/9bd3b5c2b4774d5f33b2d544f1460be9df7df2fe42f352135381c347c69a/ipython_genutils-0.2.0-py2.py3-none-any.whl\n",
    "Building wheels for collected packages: thoth-lab\n",
    "  Running setup.py bdist_wheel for thoth-lab: started\n",
    "  Running setup.py bdist_wheel for thoth-lab: finished with status 'done'\n",
    "  Stored in directory: /tmp/pip-ephem-wheel-cache-_4menpyv/wheels/c6/6e/7d/ef751ae03b7d549816c6658c7f8065cb889a8de3f0c2b3ca6c\n",
    "Successfully built thoth-lab\n",
    "selinon 0.1.0rc9 requires celery>=4, which is not installed.\n",
    "thoth-analyzer 0.1.0 requires distro, which is not installed.\n",
    "thoth-storages 0.9.5 requires aiogremlin==3.2.6rc1, which is not installed.\n",
    "tensorflow-tensorboard 1.5.1 has requirement bleach==1.5.0, but you'll have bleach 2.1.4 which is incompatible.\n",
    "tensorflow-tensorboard 1.5.1 has requirement html5lib==0.9999999, but you'll have html5lib 1.0.1 which is incompatible.\n",
    "tensorboard 1.6.0 has requirement bleach==1.5.0, but you'll have bleach 2.1.4 which is incompatible.\n",
    "tensorboard 1.6.0 has requirement html5lib==0.9999999, but you'll have html5lib 1.0.1 which is incompatible.\n",
    "notebook 5.7.0 has requirement jupyter-client>=5.2.0, but you'll have jupyter-client 5.1.0 which is incompatible.\n",
    "notebook 5.7.0 has requirement pyzmq>=17, but you'll have pyzmq 16.0.3 which is incompatible.\n",
    "kaggle 1.4.7.1 has requirement urllib3<1.23.0,>=1.15, but you'll have urllib3 1.24.1 which is incompatible.\n",
    "botocore 1.11.3 has requirement urllib3<1.24,>=1.20, but you'll have urllib3 1.24.1 which is incompatible.\n",
    "awscli 1.14.25 has requirement botocore==1.8.29, but you'll have botocore 1.11.3 which is incompatible.\n",
    "awscli 1.14.25 has requirement colorama<=0.3.7,>=0.2.5, but you'll have colorama 0.3.9 which is incompatible.\n",
    "awscli 1.14.25 has requirement PyYAML<=3.12,>=3.10, but you'll have pyyaml 3.13 which is incompatible.\n",
    "thoth-storages 0.9.5 has requirement yarl<1.2, but you'll have yarl 1.3.0 which is incompatible.\n",
    "spyder-kernels 0.2.6 has requirement jupyter-client>=5.2.3, but you'll have jupyter-client 5.1.0 which is incompatible.\n",
    "spyder-kernels 0.2.6 has requirement pyzmq>=17, but you'll have pyzmq 16.0.3 which is incompatible.\n",
    "rstcheck 2.2 has requirement sphinx<1.5,>=1.3, but you'll have sphinx 1.8.1 which is incompatible.\n",
    "gremlinpython 3.2.6 has requirement six==1.10.0, but you'll have six 1.12.0 which is incompatible.\n",
    "gremlinpython 3.2.6 has requirement tornado==4.4.1, but you'll have tornado 4.5.2 which is incompatible.\n",
    "Installing collected packages: decorator, networkx, idna, urllib3, certifi, chardet, requests, six, python-dateutil, pytz, numpy, pandas, retrying, ipython-genutils, traitlets, jupyter-core, jsonschema, nbformat, plotly, thoth-lab\n",
    "  Found existing installation: decorator 4.3.2\n",
    "    Uninstalling decorator-4.3.2:\n",
    "      Successfully uninstalled decorator-4.3.2\n",
    "  Found existing installation: networkx 2.2\n",
    "    Uninstalling networkx-2.2:\n",
    "      Successfully uninstalled networkx-2.2\n",
    "  Found existing installation: idna 2.8\n",
    "    Uninstalling idna-2.8:\n",
    "      Successfully uninstalled idna-2.8\n",
    "  Found existing installation: urllib3 1.24.1\n",
    "    Uninstalling urllib3-1.24.1:\n",
    "      Successfully uninstalled urllib3-1.24.1\n",
    "  Found existing installation: certifi 2018.11.29\n",
    "Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.\n",
    "You are using pip version 18.1, however version 19.0.2 is available.\n",
    "You should consider upgrading via the 'pip install --upgrade pip' command.\n",
    "  </pre>\n",
    "</details>\n",
    "\n",
    "In the log file, there is a line 76 which contains the information about the package which broke the installation: `certifi`. We want to extract this information. In the most common case, we want to get a **report** containing the observations about this particular build log.\n",
    "\n",
    "```python\n",
    "from thoth.build_analysers.analysis import build_breaker_report\n",
    "\n",
    "log: str = ...\n",
    "report: dict = build_breaker_report(log)\n",
    "```\n",
    "\n",
    "The result is a dictionary which can further be formatted by the convenient `build_breaker_format_report` function.\n",
    "\n",
    "```python\n",
    "from thoth.build_analysers.analysis import build_breaker_format_report\n",
    "\n",
    "print(build_breaker_format_report(report))\n",
    "```\n",
    "\n",
    "And you should see something like this:\n",
    "```\n",
    "Build breaker:\n",
    "\n",
    "    {\n",
    "        \"already_satisfied\": null,\n",
    "        \"source\": \"thoth-lab\",\n",
    "        \"target\": \"certifi\",\n",
    "        \"version_installed\": null,\n",
    "        \"version_specified\": \">=2017.4.17\"\n",
    "    }\n",
    "\n",
    "Probable reason:\n",
    "\n",
    "    76: Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.\n",
    "\n",
    "Candidates:\n",
    "\n",
    "    [\n",
    "        \"76: Cannot uninstall 'certifi'. It is a distutils installed project and thus we cannot accurately determine which files belong to it which would lead to only a partial uninstall.\"\n",
    "    ]\n",
    "\n",
    "```\n",
    "\n",
    "<br>\n",
    "\n",
    "### How it works\n",
    "\n",
    "The logic is quite simple (and also a bit naive). Really, no rocket science, just a lot of string matching.\n",
    "\n",
    "#### The intuition\n",
    "\n",
    "The approach currently handles logs produced by pip and pipenv package installers. For both of them, there is a set of patterns which, if found in a log, potentially indicate an error. These patterns are strings formatted according to the [PEP3101](https://www.python.org/dev/peps/pep-3101/) string formatting syntax and are stored in the [`resources`](/thoth/build_analysers/resources) folder as [pandas DataFrames](https://pandas.pydata.org/pandas-docs/version/0.23.4/generated/pandas.DataFrame.html).\n",
    "\n",
    "We evaluate the 'goodness' of a match for each line of the log file with respect to the relevant set of patterns and label them based on the given *thresholds*. Currently, the `ERROR` threshold is set to the score of *>=0.6* on the scale from *0* through *1.0*. Finally, we try to search for package names in the log file to construct a dependency table. We filter the `ERROR`s containing a package name and we've got ourselves a list of *build breakers*.\n",
    "\n",
    "<br>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Installation\n",
    "\n",
    "The package is available on [PyPI](https://pypi.org/project/thoth-build-analysers/):\n",
    "\n",
    "```bash\n",
    "pip install thoth-build-analysers\n",
    "```\n",
    "\n",
    "And you're good to go :thumbsup:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-14T15:47:13.720463Z",
     "start_time": "2019-05-14T15:47:13.666163Z"
    },
    "init_cell": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The autoreload extension is already loaded. To reload it, use:\n",
      "  %reload_ext autoreload\n"
     ]
    }
   ],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "init_cell": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.1.0rc0'"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from thoth.build_analysers import __version__\n",
    "\n",
    "__version__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-14T15:47:13.844581Z",
     "start_time": "2019-05-14T15:47:13.798669Z"
    },
    "init_cell": true
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "from thoth.build_analysers.analysis import build_breaker_analyze\n",
    "from thoth.build_analysers.analysis import build_breaker_report\n",
    "\n",
    "from thoth.build_analysers.analysis import build_breaker_format_report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Download fixtures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Download the source distribution which contains the fixtures (or use your own log file to analyse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Looking in indexes: https://test.pypi.org/simple/\n",
      "Collecting thoth-build-analysers\n",
      "  File was already downloaded /tmp/thoth_build_analysers-0.1.0rc0-py3-none-any.whl\n",
      "Successfully downloaded thoth-build-analysers\n"
     ]
    }
   ],
   "source": [
    "!pip download --no-deps -i https://test.pypi.org/simple/ -d /tmp thoth-build-analysers\n",
    "!tar -xzf /tmp/thoth-build-analysers-0.1.0rc0.tar.gz -C /tmp/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixtures_dir = Path(f\"/tmp/thoth-build-analysers-{__version__}/fixtures/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "PosixPath('/tmp/thoth-build-analysers-0.1.0rc0/fixtures/pipenv.failed.log')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from random import choice\n",
    "\n",
    "log_fpath = choice(list(fixtures_dir.iterdir()))\n",
    "log_fpath"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read the raw log file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'[pipenv.exceptions.ResolutionFailure]:       req_dir=requirements_dir\\n[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 726, in resolve_deps\\n[pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,\\n[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 480, in actually_resolve_deps\\n[pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()\\n[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 395, in resolve\\n[pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))\\n[pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\\n[pipenv.exceptions.ResolutionFailure]:       Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0,...'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "log = log_fpath.read_text()\n",
    "log[:1000] + \"...\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Report"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-14T15:47:13.277256Z",
     "start_time": "2019-05-14T15:47:02.831855Z"
    },
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length of the log file: 35\n",
      "\n",
      "The log comes from: pipenv\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>msg</th>\n",
       "      <th>score</th>\n",
       "      <th>pattern</th>\n",
       "      <th>label</th>\n",
       "      <th>colour</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       p...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       T...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       S...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]: Warning...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>First try clearing your dependency cache with ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>First try clearing your dependency cache with ...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Alternatively, you can use $ pipenv install --...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alternatively, you can use {} to bypass this m...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Hint: try $ pipenv lock --pre if it is a pre-r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Hint: try {} if it is a pre-release dependency.</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ERROR: ERROR: Could not find a version that ma...</td>\n",
       "      <td>0.672544</td>\n",
       "      <td>Could not find a version that matches {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#c7e77f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2....</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Tried: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Skipped pre-versions: 1.0b1</td>\n",
       "      <td>0.384311</td>\n",
       "      <td>pre-versions: {}</td>\n",
       "      <td>WARNING</td>\n",
       "      <td>#fece7c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>There are incompatible versions in the resolve...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>There are incompatible versions in the resolve...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:   File ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       p...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       T...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]:       S...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>[pipenv.exceptions.ResolutionFailure]: Warning...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>[pipenv.exceptions.{}]: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>First try clearing your dependency cache with ...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>First try clearing your dependency cache with ...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>Alternatively, you can use $ pipenv install --...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Alternatively, you can use {} to bypass this m...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Hint: try $ pipenv lock --pre if it is a pre-r...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Hint: try {} if it is a pre-release dependency.</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>ERROR: ERROR: Could not find a version that ma...</td>\n",
       "      <td>0.672544</td>\n",
       "      <td>Could not find a version that matches {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#c7e77f</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2....</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>Tried: {}</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>Skipped pre-versions: 1.0b1</td>\n",
       "      <td>0.384311</td>\n",
       "      <td>pre-versions: {}</td>\n",
       "      <td>WARNING</td>\n",
       "      <td>#fece7c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>There are incompatible versions in the resolve...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>There are incompatible versions in the resolve...</td>\n",
       "      <td>ERROR</td>\n",
       "      <td>#17934e</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                  msg     score  \\\n",
       "0   [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "1   [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "2   [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "3   [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "4   [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "5   [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "6   [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "7   [pipenv.exceptions.ResolutionFailure]:       p...  1.000000   \n",
       "8   [pipenv.exceptions.ResolutionFailure]:       T...  1.000000   \n",
       "9   [pipenv.exceptions.ResolutionFailure]:       S...  1.000000   \n",
       "10  [pipenv.exceptions.ResolutionFailure]: Warning...  1.000000   \n",
       "11  First try clearing your dependency cache with ...  1.000000   \n",
       "12  Alternatively, you can use $ pipenv install --...  1.000000   \n",
       "13  Hint: try $ pipenv lock --pre if it is a pre-r...  1.000000   \n",
       "14  ERROR: ERROR: Could not find a version that ma...  0.672544   \n",
       "15  Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2....  1.000000   \n",
       "16                        Skipped pre-versions: 1.0b1  0.384311   \n",
       "17  There are incompatible versions in the resolve...  1.000000   \n",
       "18  [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "19  [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "20  [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "21  [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "22  [pipenv.exceptions.ResolutionFailure]:   File ...  1.000000   \n",
       "23  [pipenv.exceptions.ResolutionFailure]:       r...  1.000000   \n",
       "24  [pipenv.exceptions.ResolutionFailure]:       p...  1.000000   \n",
       "25  [pipenv.exceptions.ResolutionFailure]:       T...  1.000000   \n",
       "26  [pipenv.exceptions.ResolutionFailure]:       S...  1.000000   \n",
       "27  [pipenv.exceptions.ResolutionFailure]: Warning...  1.000000   \n",
       "28  First try clearing your dependency cache with ...  1.000000   \n",
       "29  Alternatively, you can use $ pipenv install --...  1.000000   \n",
       "30  Hint: try $ pipenv lock --pre if it is a pre-r...  1.000000   \n",
       "31  ERROR: ERROR: Could not find a version that ma...  0.672544   \n",
       "32  Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2....  1.000000   \n",
       "33                        Skipped pre-versions: 1.0b1  0.384311   \n",
       "34  There are incompatible versions in the resolve...  1.000000   \n",
       "\n",
       "                                              pattern    label   colour  \n",
       "0                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "1                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "2                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "3                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "4                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "5                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "6                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "7                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "8                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "9                          [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "10                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "11  First try clearing your dependency cache with ...    ERROR  #17934e  \n",
       "12  Alternatively, you can use {} to bypass this m...    ERROR  #17934e  \n",
       "13    Hint: try {} if it is a pre-release dependency.    ERROR  #17934e  \n",
       "14           Could not find a version that matches {}    ERROR  #c7e77f  \n",
       "15                                          Tried: {}    ERROR  #17934e  \n",
       "16                                   pre-versions: {}  WARNING  #fece7c  \n",
       "17  There are incompatible versions in the resolve...    ERROR  #17934e  \n",
       "18                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "19                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "20                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "21                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "22                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "23                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "24                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "25                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "26                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "27                         [pipenv.exceptions.{}]: {}    ERROR  #17934e  \n",
       "28  First try clearing your dependency cache with ...    ERROR  #17934e  \n",
       "29  Alternatively, you can use {} to bypass this m...    ERROR  #17934e  \n",
       "30    Hint: try {} if it is a pre-release dependency.    ERROR  #17934e  \n",
       "31           Could not find a version that matches {}    ERROR  #c7e77f  \n",
       "32                                          Tried: {}    ERROR  #17934e  \n",
       "33                                   pre-versions: {}  WARNING  #fece7c  \n",
       "34  There are incompatible versions in the resolve...    ERROR  #17934e  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "handler, analysis = build_breaker_analyze(log)\n",
    "\n",
    "print(f\"\\nThe log comes from: {handler}\")\n",
    "analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-14T15:47:13.513759Z",
     "start_time": "2019-05-14T15:47:13.280359Z"
    },
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'build_breaker': {'already_satisfied': None,\n",
       "  'source': None,\n",
       "  'target': 'six',\n",
       "  'version_installed': None,\n",
       "  'version_specified': '==1.10.0,>=1.11.0,>=1.5,>=1.9.0'},\n",
       " 'reason': {'ln': 31,\n",
       "  'msg': 'ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0'},\n",
       " 'candidates': [{'ln': 34,\n",
       "   'msg': 'There are incompatible versions in the resolved dependencies.'},\n",
       "  {'ln': 32,\n",
       "   'msg': 'Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0'},\n",
       "  {'ln': 1,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 726, in resolve_deps'},\n",
       "  {'ln': 2,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,'},\n",
       "  {'ln': 3,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 480, in actually_resolve_deps'},\n",
       "  {'ln': 4,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()'},\n",
       "  {'ln': 5,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 395, in resolve'},\n",
       "  {'ln': 6,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))'},\n",
       "  {'ln': 7,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0'},\n",
       "  {'ln': 8,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0'},\n",
       "  {'ln': 9,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 1.0b1'},\n",
       "  {'ln': 10,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.'},\n",
       "  {'ln': 11,\n",
       "   'msg': 'First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.'},\n",
       "  {'ln': 12,\n",
       "   'msg': 'Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.'},\n",
       "  {'ln': 13,\n",
       "   'msg': 'Hint: try $ pipenv lock --pre if it is a pre-release dependency.'},\n",
       "  {'ln': 15,\n",
       "   'msg': 'Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0'},\n",
       "  {'ln': 17,\n",
       "   'msg': 'There are incompatible versions in the resolved dependencies.'},\n",
       "  {'ln': 18,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 726, in resolve_deps'},\n",
       "  {'ln': 19,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,'},\n",
       "  {'ln': 20,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 480, in actually_resolve_deps'},\n",
       "  {'ln': 21,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()'},\n",
       "  {'ln': 22,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:   File \"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\", line 395, in resolve'},\n",
       "  {'ln': 23,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))'},\n",
       "  {'ln': 24,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0'},\n",
       "  {'ln': 25,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0'},\n",
       "  {'ln': 26,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 1.0b1'},\n",
       "  {'ln': 27,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.'},\n",
       "  {'ln': 28,\n",
       "   'msg': 'First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.'},\n",
       "  {'ln': 29,\n",
       "   'msg': 'Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.'},\n",
       "  {'ln': 30,\n",
       "   'msg': 'Hint: try $ pipenv lock --pre if it is a pre-release dependency.'},\n",
       "  {'ln': 0,\n",
       "   'msg': '[pipenv.exceptions.ResolutionFailure]:       req_dir=requirements_dir'},\n",
       "  {'ln': 14,\n",
       "   'msg': 'ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0'},\n",
       "  {'ln': 31,\n",
       "   'msg': 'ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0'}]}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "report: dict = build_breaker_report(analysis, handler=handler)\n",
    "report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "ExecuteTime": {
     "end_time": "2019-05-14T15:47:13.594697Z",
     "start_time": "2019-05-14T15:47:13.519538Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Build breaker:\n",
      "\n",
      "    {\n",
      "        \"already_satisfied\": null,\n",
      "        \"source\": null,\n",
      "        \"target\": \"six\",\n",
      "        \"version_installed\": null,\n",
      "        \"version_specified\": \"==1.10.0,>=1.11.0,>=1.5,>=1.9.0\"\n",
      "    }\n",
      "\n",
      "Probable reason:\n",
      "\n",
      "    31: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\n",
      "\n",
      "Candidates:\n",
      "\n",
      "    [\n",
      "        \"34: There are incompatible versions in the resolved dependencies.\",\n",
      "        \"32: Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0\",\n",
      "        \"1: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 726, in resolve_deps\",\n",
      "        \"2: [pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,\",\n",
      "        \"3: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 480, in actually_resolve_deps\",\n",
      "        \"4: [pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()\",\n",
      "        \"5: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 395, in resolve\",\n",
      "        \"6: [pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))\",\n",
      "        \"7: [pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\",\n",
      "        \"8: [pipenv.exceptions.ResolutionFailure]:       Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0\",\n",
      "        \"9: [pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 1.0b1\",\n",
      "        \"10: [pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.\",\n",
      "        \"11: First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.\",\n",
      "        \"12: Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.\",\n",
      "        \"13: Hint: try $ pipenv lock --pre if it is a pre-release dependency.\",\n",
      "        \"15: Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0\",\n",
      "        \"17: There are incompatible versions in the resolved dependencies.\",\n",
      "        \"18: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 726, in resolve_deps\",\n",
      "        \"19: [pipenv.exceptions.ResolutionFailure]:       req_dir=req_dir,\",\n",
      "        \"20: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 480, in actually_resolve_deps\",\n",
      "        \"21: [pipenv.exceptions.ResolutionFailure]:       resolved_tree = resolver.resolve()\",\n",
      "        \"22: [pipenv.exceptions.ResolutionFailure]:   File \\\"/home/macermak/.local/lib/python3.6/site-packages/pipenv/utils.py\\\", line 395, in resolve\",\n",
      "        \"23: [pipenv.exceptions.ResolutionFailure]:       raise ResolutionFailure(message=str(e))\",\n",
      "        \"24: [pipenv.exceptions.ResolutionFailure]:       pipenv.exceptions.ResolutionFailure: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\",\n",
      "        \"25: [pipenv.exceptions.ResolutionFailure]:       Tried: 0.9.0, 0.9.1, 0.9.2, 1.0.0, 1.1.0, 1.2.0, 1.3.0, 1.4.0, 1.4.1, 1.5.0, 1.5.0, 1.5.1, 1.5.1, 1.5.2, 1.5.2, 1.6.0, 1.6.0, 1.6.1, 1.6.1, 1.7.0, 1.7.0, 1.7.1, 1.7.1, 1.7.2, 1.7.2, 1.7.3, 1.7.3, 1.8.0, 1.8.0, 1.9.0, 1.9.0, 1.10.0, 1.10.0, 1.11.0, 1.11.0, 1.12.0, 1.12.0\",\n",
      "        \"26: [pipenv.exceptions.ResolutionFailure]:       Skipped pre-versions: 1.0b1\",\n",
      "        \"27: [pipenv.exceptions.ResolutionFailure]: Warning: Your dependencies could not be resolved. You likely have a mismatch in your sub-dependencies.\",\n",
      "        \"28: First try clearing your dependency cache with $ pipenv lock --clear, then try the original command again.\",\n",
      "        \"29: Alternatively, you can use $ pipenv install --skip-lock to bypass this mechanism, then run $ pipenv graph to inspect the situation.\",\n",
      "        \"30: Hint: try $ pipenv lock --pre if it is a pre-release dependency.\",\n",
      "        \"0: [pipenv.exceptions.ResolutionFailure]:       req_dir=requirements_dir\",\n",
      "        \"14: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\",\n",
      "        \"31: ERROR: ERROR: Could not find a version that matches six==1.10.0,>=1.11.0,>=1.5,>=1.9.0\"\n",
      "    ]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "print(build_breaker_format_report(report))"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Initialization Cell",
  "hide_input": false,
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": true,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
